library/pkparse.c

```diff
@@ -446,6 +446,148 @@ static int pk_group_id_from_specified(const mbedtls_asn1_buf *params,
 }
 #endif /* MBEDTLS_PK_PARSE_EC_EXTENDED */
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+/*
+ * Use EC parameters to initialise an EC group
+ *
+ * ECParameters ::= CHOICE {
+ *   namedCurve         OBJECT IDENTIFIER
+ *   specifiedCurve     SpecifiedECDomain -- = SEQUENCE { ... }
+ *   -- implicitCurve   NULL
+ */
+static int pk_use_ecparams(const mbedtls_asn1_buf *params, mbedtls_pk_context *pk)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    mbedtls_ecp_group_id grp_id;
+    psa_ecc_family_t ec_family;
+    size_t bits;
+
+    if (params->tag == MBEDTLS_ASN1_OID) {
+        if (mbedtls_oid_get_ec_grp(params, &grp_id) != 0) {
+            return MBEDTLS_ERR_PK_UNKNOWN_NAMED_CURVE;
+        }
+    } else {
+#if defined(MBEDTLS_PK_PARSE_EC_EXTENDED)
+        if ((ret = pk_group_id_from_specified(params, &grp_id)) != 0) {
+            return ret;
+        }
+#else /* MBEDTLS_PK_PARSE_EC_EXTENDED */
+        return MBEDTLS_ERR_PK_KEY_INVALID_FORMAT;
+#endif /* MBEDTLS_PK_PARSE_EC_EXTENDED */
+    }
+
+    ec_family = mbedtls_ecc_group_to_psa(grp_id, &bits);
+
+    /*
+     * pk_ec_family may already be initialized; if so, make sure IDs match
+     */
+    if ((pk->pk_ec_family != 0) && (pk->pk_ec_family != ec_family)) {
+        return MBEDTLS_ERR_PK_KEY_INVALID_FORMAT;
+    }
+
+    /* This function can be called also while working on private keys, not
+     * only the public ones. Therefore we keep also the update of the
+     * "mbedtls_ecp_keypair" part for now. This can be removed once also
+     * the private key is kept as raw data on the "mbedtls_pk_context"
+     * structure */
+    if ((ret = mbedtls_ecp_group_load(&(mbedtls_pk_ec(*pk)->grp),
+                                      grp_id)) != 0) {
+        return ret;
+    }
+
+    pk->pk_ec_family = ec_family;
+    pk->pk_bits = bits;
+
+    return 0;
+}
+
+/*
+ * Create a temporary ecp_keypair for converting an EC point in compressed
+ * format to an uncompressed one
+ */
+static int pk_convert_compressed_ec(mbedtls_pk_context *pk,
+                                    const unsigned char *in_start, size_t in_len,
+                                    size_t *out_buf_len, unsigned char *out_buf,
+                                    size_t out_buf_size)
+{
+    mbedtls_ecp_keypair ecp_key;
+    mbedtls_ecp_group_id ecp_group_id;
+    int ret;
+
+    ecp_group_id = mbedtls_ecc_group_of_psa(pk->pk_ec_family, pk->pk_bits, 0);
+
+    mbedtls_ecp_keypair_init(&ecp_key);
+    ret = mbedtls_ecp_group_load(&(ecp_key.grp), ecp_group_id);
+    if (ret != 0) {
+        return ret;
+    }
+    ret = mbedtls_ecp_point_read_binary(&(ecp_key.grp), &ecp_key.Q,
+                                        in_start, in_len);
+    if (ret != 0) {
+        mbedtls_ecp_keypair_free(&ecp_key);
```

Comment:
This code can be optimized (mbedtls_ecp_keypair_free() is now duplicated on each error) as in other places by adding `exit` label at the end of function that performs:
```
mbedtls_ecp_keypair_free(&ecp_key);
return ret;
```